text/0007-service-modules.md

```diff
@@ -0,0 +1,128 @@
+
+## Django app structure
+
+We'll use 'discover' as an example for the service modules
+
+```
+src/sentry/discover
+    __init__.py
+    app.py
+    urls.py
+    models/__init__.py
+    models/discoversavedquery.py
+    endpoints/discoverquery.py
+    serializers/discoverquery.py
+    tasks/deduplicate_things.py
+
+    tests/__init__.py
+    tests/models/test_discoversavedquery.py
+    tests/endpoints/test_discoverquery.py
+    tests/serializers/test_discoverquery.py
+```
+
+In addition to the Django related modules, celery tasks, consumers and any other
+modules can be contained within a service. If a service doesn't provide
+endpoints or use models it can still benefit from the proposed structure.
+
+## Test location
+
+Tests for a service would continue to live inside the top-level `tests`
+directory. The `tests` tree would mirror continue to mirror the service + module
+structure of the application code. Sharing naming conventions should make
+running sub-sets of tests simpler to automate.
+
+## Formal entry points
+
+Service modules would use `__init__.py` to define the interface they present to
+the rest of the Sentry monolith. Having the public interface of a service
+formally defined limits the amount of entanglement the rest of the application
+can create.
+
+## Importing service internals is not allowed
+
+An important change from the present application structure is that modules
+outside of a service's scope would be *disallowed* from importing modules inside
+a service. Modules outside of a service boundary may only import the top-level
+service.
+
+Disallowing cross-service internal imports could be enforced with the
+[flake8-import-graph](https://pypi.org/project/flake8-import-graph/) extension.
+
+# Options Considered
+
+Another approach to this would be to put 'services' inside the directories of
+each 'kind'. Again using discover as an example:
+
+```
+src/sentry
+  endpoints/discover/discover_query.py
+  models/discover/discoversavedquery.py
+  tests/models/discover/test_discoversavedquery.py
+```
+
+This approach dilutes the consistency benefits, and requires a significantly
+more complex import graph rules. It also does not improve local development or
+offer benefits to CI subsetting.
+
+# Drawbacks
+
+This approach will require moving **most** of the application source code
+around. We currently store classpaths in several locations in the database. We
+may need to use data migrations to update these paths or maintain aliases for
+compatibility.
+
+Co-locating tests with application code is a potentially contentious change.
+While co-located tests are compatible with pytest test discovery works, it is an
+[un conventional](un-conventional) approach in python projects.
+
+# Unresolved questions
+
+* What 'services' would we need to add to the application?
+* What do we do with models and logic that is shared by many endpoints/domains?
+  Examples of this include rate limiting, and models like Organization, and
+  Project?
```

Comment:
There must be no "common" or "utils" or "misc" directory, or you'll eventually get a fresh new cobweb of code with unclear ownership. Under the current proposal, as written, each (kind of, or owner of) shared code must be given a name and a directory at `src/sentry/*`.

to your examples

  * src/sentry/rate-limiting
  * src/sentry/project/models/project.py
  * src/sentry/organization/models/organization.py

It's tempting to combine project and organization into a single "core" or "sentry" service, but then there's no clarity as to what should be added to (or rejected from) that service. Keep in mind that the "project" service will not contain just the one file but all the  endpoints, serializers, tasks and tests that are specific to "project". And any proposed additions to this service will have a very clear admission criteria: is it actually about "project" or does it have some other scope.

But that's just my two cents :)

Reply:
The point that I tried to make was that "core" by definition has no clear boundary. Even if you make a document about its boundary, that doc is liable to be "improved" or ignored. Whereas a "project" service does have clear boundary by definition.

Note: I'm just trying to be extra clear. I don't actually have a strong opinion on this.

Reply:
> Note: I'm just trying to be extra clear. I don't actually have a strong opinion on this.

Thank you for the clarification. I agree that 'core' will be magnet of increasing scope. 'projects' could be a good container for organization, project and related settings that those resources have.